### PR TITLE
Multi tenancy

### PR DESCRIPTION
**Pull request by Jonathan-Eng**

### CODE REVIEW

File: accounts-service/client/src/main/java/pojos/Account.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package pojos;
+
+public class Account {
+    private long id;
+    private String name;
+    private String token;
+    private String esindex;
```
Comment (naorgu) [inline accounts-service/client/src/main/java/pojos/Account.java]:
```suggestion
    private String esIndex;
```

File: accounts-service/src/main/java/config/AccountsServiceConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package config;
+
+import java.util.Map;
+
+public class AccountsServiceConfiguration {
```
Comment (naorgu) [inline accounts-service/src/main/java/config/AccountsServiceConfiguration.java]: Looks like it is unused

File: accounts-service/src/main/java/juice/modules/AccountsServiceModule.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import main.AccountsService;
+
+public class AccountsServiceModule extends AbstractModule {
+
+    public void configure() {
+        install(new ServerModule());
+        install(new AccountsDbModule());
+    }
+
+    /*
```
Comment (naorgu) [inline accounts-service/src/main/java/juice/modules/AccountsServiceModule.java]: remove

File: accounts-service/src/main/java/jettyjersey/CreateAccountBody.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package jettyjersey;
+
+public class CreateAccountBody {
```
Comment (naorgu) [inline accounts-service/src/main/java/jettyjersey/CreateAccountBody.java]:
```suggestion
public class CreateAccountRequest {
```

File: accounts-service/client/src/main/java/api/AccountsServiceApi.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                    .get();
+        }
+
+        /**
+         * Creates a new account with the account's name
+         * @param accountName
+         * @return
+         */
+        public Response createAccount(String accountName) {
+            String jsonRequestStr = String.format("{\"accountName\": \"%s\"}", accountName);
```
Comment (naorgu) [inline accounts-service/client/src/main/java/api/AccountsServiceApi.java]: You can do instead `new CreateAccountRequest(accountName)`

File: indexer/src/main/java/consumer/IndexerConsumer.java [hunk 6]
```diff
@@ -26,6 +30,7 @@
     private final RestHighLevelClient elasticsearchClient;
     private String topic = "my_topic";
     private static final Logger logger = LogManager.getLogger(IndexerConsumer.class);
+    private final AccountsServiceApi accountsServiceApi;
 
     public void setTopic(String topic) {
```
Comment (naorgu) [inline indexer/src/main/java/consumer/IndexerConsumer.java]: Setter is no beseder

File: indexer/src/main/java/consumer/IndexerConsumer.java [hunk 7]
```diff
@@ -35,6 +40,7 @@ public void setTopic(String topic) {
     public IndexerConsumer(KafkaConsumer<String, String> consumer, RestHighLevelClient elasticsearchClient) {
         this.consumer = requireNonNull(consumer);
         this.elasticsearchClient = requireNonNull(elasticsearchClient);
+        this.accountsServiceApi = new AccountsServiceApi();
```
Comment (naorgu) [inline indexer/src/main/java/consumer/IndexerConsumer.java]: This should be injected using Guice, right?

File: logs-shipper/src/main/java/jettyjersey/SearchResource.java [hunk 8]
```diff
@@ -31,43 +39,45 @@
 public class SearchResource {
 
     private final RestHighLevelClient esClient;
+    private final AccountsServiceApi accountsServiceApi;
 
     @Inject
     public SearchResource(RestHighLevelClient esClient) {
         this.esClient = requireNonNull(esClient);
+        this.accountsServiceApi = new AccountsServiceApi();
```
Comment (naorgu) [inline logs-shipper/src/main/java/jettyjersey/SearchResource.java]: should be injected

File: logs-shipper/src/main/java/jettyjersey/IndexResource.java [hunk 9]
```diff
@@ -30,10 +35,12 @@
     private static int prodRecNum = 0;
     private String topic = "my_topic";
     private final static Logger logger = LogManager.getLogger(IndexResource.class);
+    private AccountsServiceApi accountsServiceApi;
 
     @Inject
     public IndexResource(KafkaProducer<String, String> prod) {
         this.prod = requireNonNull(prod);
+        this.accountsServiceApi = new AccountsServiceApi();
```
Comment (naorgu) [inline logs-shipper/src/main/java/jettyjersey/IndexResource.java]: Should be injected

File: logs-shipper/src/main/java/jettyjersey/IndexResource.java [hunk 10]
```diff
@@ -30,10 +35,12 @@
     private static int prodRecNum = 0;
     private String topic = "my_topic";
```
Comment (naorgu) [inline logs-shipper/src/main/java/jettyjersey/IndexResource.java]: The topic should be part of the configuration

File: indexer/src/main/java/consumer/IndexerConsumer.java [hunk 11]
```diff
@@ -26,6 +30,7 @@
     private final RestHighLevelClient elasticsearchClient;
     private String topic = "my_topic";
```
Comment (naorgu) [inline indexer/src/main/java/consumer/IndexerConsumer.java]: should be configurable

File: indexer/src/main/java/consumer/IndexerConsumer.java [hunk 12]
```diff
@@ -45,34 +51,41 @@ public void consume() {
...
         }
     }
+
+    private BulkRequest getBulkRequestForRecords(ConsumerRecords<String, String> records) {
+        BulkRequest bulkRequest = new BulkRequest();
+        records.forEach(r -> {
+
+            // index record to ES
+            try {
+                Map<String, String> rValAsMap = JsonParser.getObjFromJsonString(r.value(), Map.class);
```
Comment (naorgu) [inline indexer/src/main/java/consumer/IndexerConsumer.java]: Should be `Map<String, Object>`

File: inf/src/main/java/parser/JsonParser.java [hunk 13]
```diff
@@ -8,10 +9,20 @@
...
+        try {
+            return objectMapper.readValue(jsonString, clazz);
+        } catch (IOException e) {
+            throw new RuntimeException(String.format("Could not parse String %s to object of class %s", jsonString, clazz.getName()));
+        }
     }
-    public static <T> String getJsonStringFromObject(T obj) throws IOException {
-        return objectMapper.writeValueAsString(obj);
+
+    public static <T> String getJsonStringFromObject(T obj) {
```
Comment (naorgu) [inline inf/src/main/java/parser/JsonParser.java]: Notice here the generic `<T>` has no use - could just accept `Object` instead

File: logs-shipper/src/main/java/jettyjersey/IndexResource.java [hunk 14]
```diff
@@ -30,10 +35,12 @@
...
     private final static Logger logger = LogManager.getLogger(IndexResource.class);
+    private AccountsServiceApi accountsServiceApi;
 
     @Inject
     public IndexResource(KafkaProducer<String, String> prod) {
         this.prod = requireNonNull(prod);
+        this.accountsServiceApi = new AccountsServiceApi();
     }
 
     public String getTopic() {
```
Comment (naorgu) [inline logs-shipper/src/main/java/jettyjersey/IndexResource.java]: Remove this and `setTopic`

File: indexer/src/main/java/consumer/IndexerConsumer.java [hunk 15]
```diff
@@ -45,34 +51,41 @@ public void consume() {
...
+
+                Account account = accountsServiceApi.getAccount(r.key());
+                IndexRequest indexRequest = new IndexRequest(account.getEsindex(), "_doc")
+                        .source(rValAsMap);
+                bulkRequest.add(indexRequest);
+
+                logger.debug("Record was added to bulk request: ({}, {}, {}, {})\n", r.key(), r.value(), r.partition(), r.offset());
+            } catch (Exception | AccountInvalidTokenException e) {
+                e.printStackTrace();
```
Comment (naorgu) [inline indexer/src/main/java/consumer/IndexerConsumer.java:85]: Looks like you swallow an exception here - I guess you want to return to the client some kind of error response

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java [hunk 16]
```diff
@@ -17,48 +24,115 @@
     private static final String CONFIG_FILE_NAME = "server.config";
     private static final int RANDOM_STRING_LEN = 20;
     private static final int MAX_SEC_TO_WAIT = 15;
+    private static final int ACCOUNT_RAND_LOW = 1;
+    private static final int ACCOUNT_RAND_HIGH = 10000;
+
+
+
     private final ServerConfiguration sc;
```
Comment (naorgu) [inline logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: This is unused. delete

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java [hunk 17]
```diff
@@ -17,48 +24,115 @@
...
-    public LogsShipperClient handler;
+    private LogsShipperClient handler;
+    private AccountsServiceApi accountServiceApi;
+
 
     public ResourceTest() {
         sc = ConfigurationFactory.create(
                 ConfigFileFinder.findRealPath(CONFIG_FILE_NAME),
                 ServerConfiguration.class);
```
Comment (naorgu) [inline logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: remove

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java [hunk 18]
```diff
@@ -17,48 +24,115 @@
...
     private static final int RANDOM_STRING_LEN = 20;
     private static final int MAX_SEC_TO_WAIT = 15;
+    private static final int ACCOUNT_RAND_LOW = 1;
+    private static final int ACCOUNT_RAND_HIGH = 10000;
+
+
+
     private final ServerConfiguration sc;
-    public LogsShipperClient handler;
+    private LogsShipperClient handler;
```
Comment (naorgu) [inline logs-shipper/src/test/java/jettyjersey/ResourceTest.java]:
```suggestion
    private LogsShipperClient logsShipperClient;
```

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java [hunk 19]
```diff
@@ -17,48 +24,115 @@
...
 
         boolean isMessageIndexed = response.getStatus() == HttpURLConnection.HTTP_OK;
         boolean isMessageFoundOnce = 1 == dr.getResponse().size();
 
-        return isMessageIndexed && isMessageFoundOnce;
+        assertTrue(isMessageIndexed);
+        assertTrue(isMessageFoundOnce);
+    }
+
+    private static int genRandAccountNum() {
```
Comment (naorgu) [inline logs-shipper/src/test/java/jettyjersey/ResourceTest.java]:
```suggestion
    private static int genRandAccountName() {
```

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java [hunk 20]
```diff
@@ -17,48 +24,115 @@
...
+        }
+        catch (AccountNameAlreadyExistsException e) {
+            fail("test failed. Account name already exists");
+        }
+        catch (AccountInvalidNameException e) {
+            fail("test failed. Account name is invalid");
+        }
+        throw new RuntimeException("test failed due to unexpected exception.");
     }
 }
```
Comment (naorgu) [inline logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: Let's add tests for unauthorized token when searching/indexing

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java [hunk 21]
```diff
@@ -17,48 +24,115 @@
...
+    private static int genRandAccountNum() {
+        return new Random().nextInt(ACCOUNT_RAND_HIGH - ACCOUNT_RAND_LOW) + ACCOUNT_RAND_LOW;
+    }
+
+    private static String genRandomString() {
+        return RandomStringUtils.randomAlphabetic(RANDOM_STRING_LEN);
+    }
+
+    private static String toJsonMsg(String msg) {
+        return "{\"message\":\"" + msg + "\"}";
```
Comment (naorgu) [inline logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: Let's use instead `JsonParser.getJsonStringFromObject(new IndexBody(msg))`

File: accounts-service/client/src/main/java/api/AccountsServiceApi.java [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+            return webTarget.path(CREATE_ACCOUNT_ENDPOINT)
+                    .request(MediaType.APPLICATION_JSON)
+                    .post(Entity.json(jsonRequestStr));
+        }
+    }
+
+}
```
Comment (naorgu) [inline accounts-service/client/src/main/java/api/AccountsServiceApi.java:99]: Very nice implementation

**Review by naorgu — CHANGES_REQUESTED** [review]
Kivid of a work

File: accounts-service/src/main/java/config/AccountsServiceConfiguration.java (hunk 2 above)
Comment (Jonathan-Eng) [reply accounts-service/src/main/java/config/AccountsServiceConfiguration.java]: removed

File: accounts-service/src/main/java/juice/modules/AccountsServiceModule.java (hunk 3 above)
Comment (Jonathan-Eng) [reply accounts-service/src/main/java/juice/modules/AccountsServiceModule.java]: done

File: accounts-service/src/main/java/jettyjersey/CreateAccountBody.java (hunk 4 above)
Comment (Jonathan-Eng) [reply accounts-service/src/main/java/jettyjersey/CreateAccountBody.java]: done

File: accounts-service/client/src/main/java/api/AccountsServiceApi.java (hunk 5 above)
Comment (Jonathan-Eng) [reply accounts-service/client/src/main/java/api/AccountsServiceApi.java]: done

File: indexer/src/main/java/consumer/IndexerConsumer.java (hunk 6 above)
Comment (Jonathan-Eng) [reply indexer/src/main/java/consumer/IndexerConsumer.java]: this is because you told me we should make the topic configurable.

Comment (Jonathan-Eng) [reply indexer/src/main/java/consumer/IndexerConsumer.java]: i will make it configurable with a file

File: indexer/src/main/java/consumer/IndexerConsumer.java (hunk 7 above)
Comment (Jonathan-Eng) [reply indexer/src/main/java/consumer/IndexerConsumer.java]: yes. done

File: logs-shipper/src/main/java/jettyjersey/SearchResource.java (hunk 8 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/main/java/jettyjersey/SearchResource.java]: done

File: logs-shipper/src/main/java/jettyjersey/IndexResource.java (hunk 9 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/main/java/jettyjersey/IndexResource.java]: done

File: indexer/src/main/java/consumer/IndexerConsumer.java (hunk 12 above)
Comment (Jonathan-Eng) [reply indexer/src/main/java/consumer/IndexerConsumer.java]: done

File: inf/src/main/java/parser/JsonParser.java (hunk 13 above)
Comment (Jonathan-Eng) [reply inf/src/main/java/parser/JsonParser.java]: done

File: indexer/src/main/java/consumer/IndexerConsumer.java (hunk 15 above)
Comment (Jonathan-Eng) [reply indexer/src/main/java/consumer/IndexerConsumer.java:85]: which kind of error response? throwing exceptions invoked in the foreach loop is apparently forbidden.

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java (hunk 16 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: used here :
`private static int genRandAccountNum() {
        return new Random().nextInt(ACCOUNT_RAND_HIGH - ACCOUNT_RAND_LOW) + ACCOUNT_RAND_LOW;
    }`

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java (hunk 18 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: done

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java (hunk 19 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: done

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java (hunk 21 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: done

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java (hunk 20 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: done

File: logs-shipper/src/test/java/jettyjersey/ResourceTest.java (hunk 17 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/test/java/jettyjersey/ResourceTest.java]: done

File: accounts-service/client/src/main/java/pojos/Account.java (hunk 1 above)
Comment (Jonathan-Eng) [reply accounts-service/client/src/main/java/pojos/Account.java]: done

File: indexer/src/main/java/consumer/IndexerConsumer.java (hunk 6 above)
Comment (Jonathan-Eng) [reply indexer/src/main/java/consumer/IndexerConsumer.java]: done

File: logs-shipper/src/main/java/jettyjersey/IndexResource.java (hunk 10 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/main/java/jettyjersey/IndexResource.java]: done

File: indexer/src/main/java/consumer/IndexerConsumer.java (hunk 11 above)
Comment (Jonathan-Eng) [reply indexer/src/main/java/consumer/IndexerConsumer.java]: done

File: logs-shipper/src/main/java/jettyjersey/IndexResource.java (hunk 14 above)
Comment (Jonathan-Eng) [reply logs-shipper/src/main/java/jettyjersey/IndexResource.java]: done

**Pull request merged**